seis/ms/header.go

```diff
@@ -0,0 +1,359 @@
+	}
+
+	return start
+}
+
+// SampleRate returns the decoded header sampling rate in samples per second.
+func (h RecordHeader) SampleRate() float64 {
+	return sampleRate(int(h.SampleRateFactor), int(h.SampleRateMultiplier))
+}
+
+// SampleCount returns the number of samples in the record, independent of whether they are decoded or not.
+func (h RecordHeader) SampleCount() int {
+	return int(h.NumberOfSamples)
+}
+
+// SamplePeriod converts the sample rate into a time interval, or zero.
+func (h RecordHeader) SamplePeriod() time.Duration {
+	return samplePeriod(int(h.SampleRateFactor), int(h.SampleRateMultiplier))
+}
+
+// IsValid performs a simple consistency check of the RecordHeader contents.
+func (h RecordHeader) IsValid() bool {
+	for _, b := range h.SequenceNumber {
+		if !((b >= '0' && b <= '9') || (b == ' ') || (b == 0)) {
+			return false
+		}
+	}
+
+	if q := h.DataQualityIndicator; !(q == 'D' || q == 'R' || q == 'M' || q == 'Q') {
+		return false
+	}
+
+	if b := h.ReservedByte; !(b == ' ' || b == 0) {
+		return false
+	}
+
+	if s := h.RecordStartTime; !(s.Hour <= 23 && s.Minute <= 59 && s.Second <= 60) {
+		return false
+	}
+
+	return true
+}
+
+//TODO: needs v1.15
+/*
+// Hash can be used for deduping record blocks
+func (h RecordHeader) Hash() string {
+	var hm maphash.Hash
+
+	hm.Write(h.NetworkIdentifier[:])
+	hm.Write(h.StationIdentifier[:])
+	hm.Write(h.LocationIdentifier[:])
+	hm.Write(h.ChannelIdentifier[:])
+	hm.Write([]byte{h.DataQualityIndicator})
+	hm.Write(EncodeBTime(h.RecordStartTime))
+
+	var parts [6]byte
+	binary.BigEndian.PutUint16(parts[0:2], h.NumberOfSamples)
+	binary.BigEndian.PutUint16(parts[2:4], uint16(h.SampleRateFactor))
+	binary.BigEndian.PutUint16(parts[4:6], uint16(h.SampleRateMultiplier))
+	hm.Write(parts[:])
+
+	return fmt.Sprintf("%#x", hm.Sum64())
+}
+*/
+
+// Less can be used for sorting record blocks.
+func (h RecordHeader) Less(hdr RecordHeader) bool {
+	switch {
+	case h.Network() < hdr.Network():
+		return true
+	case h.Network() > hdr.Network():
+		return false
+	case h.Station() < hdr.Station():
+		return true
+	case h.Station() > hdr.Station():
+		return false
+	case h.Location() < hdr.Location():
+		return true
+	case h.Location() > hdr.Location():
+		return false
+	case h.Channel() < hdr.Channel():
+		return true
+	case h.Channel() > hdr.Channel():
+		return false
+	case h.DataQualityIndicator < hdr.DataQualityIndicator:
+		return true
+	case h.DataQualityIndicator > hdr.DataQualityIndicator:
+		return false
+	case h.StartTime().Before(hdr.StartTime()):
+		return true
+	case h.StartTime().After(hdr.StartTime()):
+		return false
+	case h.SampleRate() < hdr.SampleRate():
+		return true
+	default:
+		return false
+	}
+}
+
+// DecodeRecordHeader returns a RecordHeader from a byte slice.
+func DecodeRecordHeader(data []byte) RecordHeader {
+	var h [RecordHeaderSize]byte
+
+	copy(h[:], data)
+
+	return RecordHeader{
+		SequenceNumber: func() [6]byte {
```

Comment:
what about `SequenceNumber:  h[0:6]` ?

Reply:
problem is h[0:6] is a byte slice but Sequence number is a [6]byte array.  It needs the copy, i.e. change:

```golang

        return RecordHeader{
                SequenceNumber: h[0:6],
                /*
                        SequenceNumber: func() [6]byte {
                                var b [6]byte
                                copy(b[:], h[0:6])
                                return b
                        }(),
                */

                DataQualityIndicator: h[6],
```

gives:

```
./header.go:230:3: cannot use h[0:6] (type []byte) as type [6]byte in field value
```

which means a function somewhere, this mechanism at least keeps the copy close and there are a lot of variations that follow.